daemon/control_ng.c

```diff
@@ -335,16 +336,15 @@ static void control_ng_incoming(struct obj *obj, str *buf, const endpoint_t *sin
 	}
 
 send_only:
-	iovlen = 3;
 
-	iov[0].iov_base = cookie.s;
-	iov[0].iov_len = cookie.len;
-	iov[1].iov_base = " ";
-	iov[1].iov_len = 1;
-	iov[2].iov_base = to_send->s;
-	iov[2].iov_len = to_send->len;
-
-	socket_sendiov(ul, iov, iovlen, sin);
+	total_len = snprintf(send_buf, sizeof(send_buf), "%.*s %.*s", cookie.len, cookie.s, to_send->len, to_send->s);
```

Comment:
I'd assume the `snprintf()` switch here is going to be substantially slower than the iov send, But perhaps this is not performance critical? (I don't know). Maybe if cookie is not big, then only compressing to_send would be better, and then we just need to set `iov[2]` to the original or compressed buffer depending on whether we are compressing?

Reply:
I guess it depends on what can appear in the uncompressed to_send buffer. Otherwise it would be a matter of a conditional I think? Something like the following pseudo-code:

```
  setup iov[0] from cookie
  setup iov[1] from " "

  if (compress(in, out) == OK) {
    setup iov[2] from out
  } else {
    setup iov[2] from in
  }
```

of course that means the receiving end needs to handle the decompression after parsing the message instead of just on  ingestion. I'll leave @rfuchs to decide on this.

Reply:
I don't have that strong opinion about this. But I was thinking doing like this will make it easier to do handle(uncompress) the message for the other third party.(because this one was easier to me when I fix it the Kamailio's RTPEngine module).

I'm just waiting for the decision.

Reply:
I haven't looked at the Kamailio portion of this yet, but generally I would agree that this extra `snprintf` could/should be avoided. You could leave the message format as it is (cookie plus space plus compressed payload), or even use the bencode facility to carry the compressed data, e.g. the message would still be cookie plus space plus bencode dictionary, only that the dictionary would be something like `{ 'zlib': '... raw data ...' }` with the compressed data then decompressing to the actual bencode dictionary. The latter could be easily implemented by extended the iovec array by a few fields and hardcoding its format thanks to the simplicity of bencode. This would also address my concerns about `is_compressed()` as well as saving the daemon from having to uncompress first in order to detect and handle retransmits.